src/reader.rs

```diff
@@ -4855,10 +4884,22 @@
             complete(&command_line[..would_be_cursor], complete_flags, &ctx);
 
         let suggestion = if completions.is_empty() {
+            // If there are no completions to suggest, fall back to icase history.
+            if let Some(result) = icase_history_result {
+                return result;
+            }
             WString::new()
         } else {
             sort_and_prioritize(&mut completions, complete_flags);
             let comp = &completions[0];
+
+            // Prefer icase history over smartcase/icase completions.
+            if let (Some(result), CaseSensitivity::Smart | CaseSensitivity::Insensitive) =
```

Comment:
This looks good overall, but I'm wondering about one case:
what if the history result is CaseSensitivity::Insensitive (e.g. FOO matching fOObar)
and the completion result is CaseSensitivity::Smart (e.g. foo matching the same)?
Do we really want to let history win, even though it's less accurate?

Reply:
I considered this, but I can't think of a case where this is actually a meaningful difference *for a given query.*
Either the query is all-lowercase (and all non-samecase matches could be considered `Smart`), or it isn't (and all non-samecase matches could be considered `Insensitive`).